dolphinscheduler-task-plugin/dolphinscheduler-task-jupyter/src/main/java/org/apache/dolphinscheduler/plugin/task/jupyter/JupyterConstants.java

```diff
@@ -33,6 +33,18 @@ private JupyterConstants() {
      */
     public static final String CONDA_ACTIVATE = "conda activate";
 
+    /**
+     * create and activate conda env from tar
+     */
+    public static final String CREATE_ENV_FROM_TAR = "mkdir jupyter_env && " +
+            "tar -xzf %s -C jupyter_env && " +
+            "source jupyter_env/bin/activate";
```

Comment:
Is it better to let users to define this path if the user's path is different from this?

Reply:
Currently, `env tar` will be downloaded from `resource center` before task execution and removed when task execution completed. https://github.com/apache/dolphinscheduler/blob/f90f0f8490d8f4cd87d2368cd0264a4837918c7d/dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java#L141-L145
https://github.com/apache/dolphinscheduler/blob/f90f0f8490d8f4cd87d2368cd0264a4837918c7d/dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java#L199-L203

However, I think it is a good idea to give a choice to define the path and install the env on workers without getting removed after task execution. In this way, the same env will not need to be downloaded every time.

WDYT @SbloodyS @zhongjiajie

Reply:
It's better. But what i mean is if the user's the compressed tar package path after decompression is inconsistent with this hard code path.

Reply:
> It's better. But what i mean is if the user's the compressed tar package path after decompression is inconsistent with this hard code path.

Seems if users use `conda pack` to pack their environment, when they run `tar` they will get the same directory structure.
![image](https://user-images.githubusercontent.com/34905992/172807688-59cbffe0-8185-437f-b831-00a557b01701.png)

Reply:
I'm not quite sure about it. Could you take a look? @zhongjiajie

Reply:
> > It's better. But what i mean is if the user's the compressed tar package path after decompression is inconsistent with this hard code path.
> 
> Seems if users use `conda pack` to pack their environment, when they run `tar` they will get the same directory structure. ![image](https://user-images.githubusercontent.com/34905992/172807688-59cbffe0-8185-437f-b831-00a557b01701.png)

I test locally, and it is :+1:. but should we add some docs to tell use run `conda pack` to create the conda tarball?

Reply:
> However, I think it is a good idea to give a choice to define the path and install the env on workers without getting removed after task execution. In this way, the same env will not need to be downloaded every time.

but I do not think so, I think they can use shell task to do that. It is odd if we do not remove resources after the task execute finish, using two cases:

* we only support not deleting resources after task is done: although the worker has already installed the env, but we still have resource configurated in task definition, So it means worker will re-install the conda env to worker. Or maybe users have to change the task definition and remove the resource after the first time they ran the task.(it is odd)
* We have given users an option about resources install and delete. then have to change the option after they ran the task first time too.

Reply:
> > However, I think it is a good idea to give a choice to define the path and install the env on workers without getting removed after task execution. In this way, the same env will not need to be downloaded every time.
> 
> but I do not think so, I think they can use shell task to do that. It is odd if we do not remove resources after the task execute finish, using two cases:
> 
> * we only support not deleting resources after task is done: although the worker has already installed the env, but we still have resource configurated in task definition, So it means worker will re-install the conda env to worker. Or maybe users have to change the task definition and remove the resource after the first time they ran the task.(it is odd)
> * We have given users an option about resources install and delete. then have to change the option after they ran the task first time too.

Alright. This makes sense to me. Since the tasks are usually triggered multiple times, considering persisting the env could be complicated and might cause unexpected issues. If users want to persist the env, it is better to use a shell task and only trigger it once at the very beginning.